include/vinecopulib/vinecop/implementation/class.ipp

```diff
@@ -657,46 +732,72 @@ Vinecop::pdf(const Eigen::MatrixXd& u, const size_t num_threads) const
     needed_hfunc1 = vine_struct_.get_needed_hfunc1();
     needed_hfunc2 = vine_struct_.get_needed_hfunc2();
   }
+  auto disc_cols = tools_select::get_disc_cols(var_types_);
 
   // initial value must be 1.0 for multiplication
-  Eigen::VectorXd vine_density = Eigen::VectorXd::Constant(u.rows(), 1.0);
+  Eigen::VectorXd pdf = Eigen::VectorXd::Constant(u.rows(), 1.0);
 
   auto do_batch = [&](const tools_batch::Batch& b) {
-    // temporary storage objects for h-functions
+    // temporary storage objects (all data must be in (0, 1))
     Eigen::MatrixXd hfunc1(b.size, d);
-    Eigen::MatrixXd hfunc2(b.size, d);
-    Eigen::MatrixXd u_e(b.size, 2);
+    hfunc1.setZero();
+    Eigen::MatrixXd hfunc2 = hfunc1;
+    Eigen::MatrixXd hfunc1_sub = hfunc1;
+    Eigen::MatrixXd hfunc2_sub = hfunc1;
+    Eigen::MatrixXd u_e(b.size, 4);
+    Eigen::MatrixXd u_sub(b.size, 4);
 
     // fill first row of hfunc2 matrix with evaluation points;
     // points have to be reordered to correspond to natural order
-    for (size_t j = 0; j < d; ++j)
+
+    for (size_t j = 0; j < d; ++j) {
       hfunc2.col(j) = u.block(b.begin, order[j] - 1, b.size, 1);
+      if (var_types_[order[j] - 1] == "d") {
+        hfunc2_sub.col(j) =
+          u.block(b.begin, d_ + disc_cols[order[j] - 1], b.size, 1);
```

Comment:
Sorry, I don't get it, why are we using data at u - 1 for the discrete variables here?

Reply:
Sorry, don't understand your question.

Reply:
For `hfunc2_sub.col(j)`, you use the column `d_ + disc_cols[order[j] - 1]` of `u` if the variable is discrete, and the column `order[j] - 1` if the variable is continuous. I get why you actually need `d_ + disc_cols[order[j] - 1]` for the discrete, but why `order[j] - 1` for continuous?

Reply:
Ah yeah, that's not necessary (since I initialize to 0 above). I'll remove it.